### ISSUE TITLE
Optimize the case when `_version` is never specified / always 0

### ISSUE THREAD

**Issue by mikemccand** [elastic/elasticsearch#19913]
@rmuir had this clever idea: because of how we compress things, Lucene's codec knows when a given doc values field is always constant, and we could use this to know that `_version` is always 0, and do sizable optimizations like skip the per-doc `_uid` lookup during indexing, don't record version in the live version map (though we still use it for e.g. real-time get, see #19787 and #19813).

This could be a nice solution because it wouldn't require any configuration on the user's part to state that they will not use optimistic concurrency features: the engine would just notice and be more efficient in this very common case.  And if users do use optimistic concurrency features, it would be slower.

But one problem here is the engine automatically increments the version each time a `_uid` is re-indexed: can we stop doing this auto-increment?

**Comment by rmuir** [comment]
To do this cleaner (assuming semantics are fixed), we should leave version empty instead of putting 0's in there. That is done for some crazy reason involving nested docs today, and should be cleaned up.

Then things are obvious: if it is not in use, it does not exist in fieldinfos at all. truly empty instead of 0.

**Comment by jimferenczi** [comment]
> But one problem here is the engine automatically increments the version each time a _uid is re-indexed: can we stop doing this auto-increment?

Why is it a problem ? For append only indices this should never happen ? The auto-increment is an optimization since we don't check if the document has changed or not when we increment the version.

**Comment by rmuir** [comment]
It is a problem because it requires an expensive PK lookup to see if the document already exists.

For the append only use case, this is a complete waste of time.

**Comment by jimferenczi** [comment]
> It is a problem because it requires an expensive PK lookup to see if the document already exists.

The fact that we check if the document already exists is another problem, no ? I might be missing something but I don't see why we could bypass this check if the _version field is constant among the segments. The fact that all documents in the index are in their initial version does not ensure that there will be no updates in the future.

**Comment by rmuir** [comment]
There are two PK checks:

1) the PK lookup + version check done by elasticsearch
2) the pk lookup check done by indexwriter (since updateDocument was called)

Both are unnecessary in an append-only case. This is only about the first one though. If the semantics are we drop updates to a document with version X when, the document already exists with version Y and Y > X, then we don't need to do a bunch of individual, expensive, per-document lookups if there is not a single document in the *entire index* with any version > X.

**Comment by jimferenczi** [comment]
Ah right, two PK checks. Thanks for the explanation.

**Comment by rmuir** [comment]
Yes the first one i think is the most costly, real random access, also requires tracking uncommitted PKs with concurrent hash maps, and so on.

The second one is batched by indexwriter and then a bunch of ids done in big sequential passes, doesn't require the use of additional memory (version maps): i think its the lesser evil, we should worry about it separately.

**Comment by jpountz** [comment]
For the record, Elasticsearch calls addDocument rather than updateDocument when the version lookup returns `NOT_FOUND` so in the append-only case there would be a single PK lookup per document (but the costly one). I believe we could replace it with `updateDocument` by just using a version number that is guaranteed to be greater than the current version of the document (eg. by allocating versions using an AtomicLong)  but this would mean the `_version` would use disk space and we would not be able to know whether the document was created or updated (which is something we return in the response today unfortunately).

**Comment by rmuir** [comment]
I don't understand why we need to store a version at all. Whenever I inspect elasticsearch indexes today, none of the `_version` fields have any actual values!

**Comment by rmuir** [comment]
> and we would not be able to know whether the document was created or updated (which is something we return in the response today unfortunately).

it is frustrating to read this when making these discussions. It comes across as "its always been this way, therefore cannot be changed". I don't think its intentional, it just comes across as "your idea cannot work because we return X in the response, because we need to check Y consistency thing, because of Z", when I am trying to say, all of X, Y, Z, I think its all broken.

When I look at the `_version` field of any elasticsearch index i ever see, it is always unused (zeros). So I **know** its possible for it to function without these stuff, since its literally storing a whole whopping amount of nothing in all of these cases. I dont want to hear "we would need disk space" because I know that is not needed, its not happening today and things work!

In other words, I don't think we need to worry about X, Y, or Z at all, lets change that stuff. I'm making the statement that the current semantics are inefficient. Its frustrating to see the indexing speeds when lucene is a lot more efficient than this and its only a matter of time before someone else catches on. I think the semantics should be changed.

**Comment by jpountz** [comment]
>  It comes across as "its always been this way, therefore cannot be changed"

Sorry, this is not what I meant. On the contrary I'm all for removing all these features that are in the way of improving indexing speed: let's remove configurable version types, information about whether the document was created of updated in the response, etc. I mentioned the latter since it is about the response format of a commonly used API, which means we need to be careful, but this is no excuse to not do the change. The fact that it is in the way of indexing speed improvements means that we will want to fix it at some point, so the sooner, the better.

**Comment by nik9000** [comment]
> I mentioned the latter since it is about the response format of a commonly used API, which means we need to be careful, but this is no excuse to not do the change

We can break the format in 6.0 if we want. We'll branch for that before long.

If we're doing the append only case we could probably just return `_version` as `0` in the response anyway, right? So maybe we don't have to change the API if we don't want to?

**Comment by bleskes** [comment]
catching up, I think there some things to clarify:

1) The *main* use case for `_version` now is to deal with out of order delivery on replicas. This is relevant for concurrent changes to the same document (which doesn't happen for the append only case) but also when we recover a shard (replayed operation from the primary translog are concurrent to index requests). 
2) Now that we are going to have #20211 , we are actually not using the version for lookups but we still store it (more on removing it in the following points).
3) Long term, the plan is to move to use seq# instead of versioning both for out of order delivery and optimistic locking. At that point my idea is to turn off the per document versioning and allow people to enable it if they want it (external versioning is the main use case here).
4) Our plan is allow people to use an index level setting to indicate that an index is append only. We also said we want people to be able to change it on a close index (this will be nice where people want a fast re-index where they know they are only indexing things once, but then enable the index for full operations). If we allow this change, we can't remove the version field completely and I think it should be populated with 1 (the first version value).  

Given (3) is the long term plan and with #20211 , I'm not sure it's worth while going through the cycles of disabling version now.

**Comment by dnhatn** [comment]
Closing this as we now have implemented seq# and append-only.
